### PR TITLE
Make browserable C++ source code into HTMLs

### PR DESCRIPTION
**Pull request by wangkuiyi**
This PR has been updated to fix https://github.com/PaddlePaddle/Paddle/issues/732.

The intention of this PR is to add an additional type of *documentation*.  When I am learning Paddle, I do need this document. Also, a new user asked the same question in https://github.com/PaddlePaddle/Paddle/issues/715.  I feel that before Paddle API is fully stable, this document would be useful for many users.

What I did is to edit our Dockerfile so that

1. make cmake records all build commands into a JSON file, and
1. install woboq code browser which follows the JSON file to convert all .cc/.h files into .html for convenient code browsing.

This idea was from my earlier work at https://github.com/wangkuiyi/paddle-code-browse.

If this PR is merged, we can run a Nginx container to serve browsable .html files as the following:

```
# load documents in a container paddle-cpu-doc
docker run -d --name paddle-cpu-doc paddle:cpu 

# run nginx to serve documents in paddle-cpu-doc
docker run -d --volumes-from paddle-cpu-doc -p 8088:80 nginx 
```

### CODE REVIEW

**Comment by reyoung** [comment]
I will try this image on my laptop first. It may spend some time for compiling. If there is no explicit error, this PR will be approved and merged then.

**Comment by wangkuiyi** [comment]
@reyoung I tried to build both CPU-only and GPU on my workstation at home. It works. I did notice that the building takes more time than before, but I didn't record the build time.  As you are going to re-do the test, can you record the build time? I am open to give up this PR or move it to my personal repo if the build time or the image size increases significantly.

File: paddle/scripts/docker/Dockerfile.gpu [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
     libgoogle-glog-dev libgflags-dev libatlas-dev libatlas3-base g++ m4 python-pip \
     python-protobuf python-numpy python-dev swig openssh-server \
     wget unzip python-matplotlib tar xz-utils bzip2 gzip coreutils \
-    sed grep graphviz libjpeg-dev zlib1g-dev doxygen && \
-    apt-get clean -y
+    sed grep graphviz libjpeg-dev zlib1g-dev doxygen \
+    clang-3.8 llvm-3.8 libclang-3.8-dev \
+    && apt-get clean -y
 RUN pip install BeautifulSoup docopt PyYAML pillow \
     'sphinx>=1.4.0' sphinx_rtd_theme breathe recommonmark
```
Comment (reyoung) [inline paddle/scripts/docker/Dockerfile.gpu]: 改成 
```
RUN pip install -U BeautifulSoup docopt PyYAML pillow \
   sphinx sphinx_rtd_theme breathe recommonmark
```

都是用这些包的最新版本就好了。主要是似乎原来的安装方式`sphinx`会有一些问题，导致paddle编译文档出错。

File: paddle/scripts/docker/Dockerfile.gpu [hunk 2]
```diff
@@ -18,6 +19,7 @@ ENV WITH_GPU=ON
 RUN mkdir /paddle
 COPY . /paddle/
 RUN /paddle/paddle/scripts/docker/build.sh
+VOLUME ["/usr/share/nginx/html/data", "/usr/share/nginx/html/paddle"]
 
 RUN echo 'export LD_LIBRARY_PATH=/usr/lib64:${LD_LIBRARY_PATH}' >> /etc/profile
 RUN pip install /usr/local/opt/paddle/share/wheels/*.whl
```
Comment (reyoung) [inline paddle/scripts/docker/Dockerfile.gpu]: 这个，似乎在macos上面用docker，编译出来的wheel包，会导致安装失败。。不过这个看起来也不是paddle的问题，似乎是pip，wheel包和docker配合有问题。。我回头再查一下，这个目前先不管了。

Comment (wangkuiyi) [inline paddle/scripts/docker/Dockerfile.gpu]: 我在linux下build一切ok。一会儿我在mac下build一下试试。

Comment (reyoung) [inline paddle/scripts/docker/Dockerfile.gpu]: 这个倒是不太重要，基本上mac上build docker image也不算是一个重要需求。

Comment (wangkuiyi) [inline paddle/scripts/docker/Dockerfile.gpu]: 我在mac上build也碰到同样的问题了。我估计之前没有在Linux上发现问题，是因为我之前在Linux上build过，所以RUN apt-get install生成的那些层被cache在我的Linux workstation上了。我正在按照你的建议修改。

File: paddle/scripts/docker/Dockerfile.gpu (hunk 1 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/Dockerfile.gpu]: Done

**Comment by wangkuiyi** [comment]
I updated this PR following @reyoung 's suggestion in the [comment](https://github.com/PaddlePaddle/Paddle/pull/723#discussion_r90824997), and tested it using `--no-cache` as mentioned in [comment](https://github.com/PaddlePaddle/Paddle/pull/723#discussion_r90904161). Both of the following commands work now:

```
docker build --no-cache -t paddle:cpu-doc -f paddle/scripts/docker/Dockerfile .
docker build --no-cache -t paddle:gpu-doc -f paddle/scripts/docker/Dockerfile.gpu .
```

File: paddle/scripts/docker/Dockerfile.gpu (hunk 2 above)
Comment (wangkuiyi) [inline paddle/scripts/docker/Dockerfile.gpu]: Done

**Pull request merged**